internal/dag/httpproxy_processor_test.go

```diff
@@ -131,3 +135,160 @@ func TestEnforceRoute(t *testing.T) {
 	}
 }
+
+func TestToCORSPolicy(t *testing.T) {
+	tests := map[string]struct {
+		cp      *contour_api_v1.CORSPolicy
+		want    *CORSPolicy
+		wantErr bool
+	}{
+		"no CORS policy": {
+			cp:   nil,
+			want: nil,
+		},
+		"all fields present and valid": {
+			cp: &contour_api_v1.CORSPolicy{
+				AllowCredentials: true,
+				AllowHeaders:     []contour_api_v1.CORSHeaderValue{"X-Some-Header-A", "X-Some-Header-B"},
+				AllowMethods:     []contour_api_v1.CORSHeaderValue{"GET", "PUT"},
+				AllowOrigin:      []string{"*"},
+				ExposeHeaders:    []contour_api_v1.CORSHeaderValue{"X-Expose-A", "X-Expose-B"},
+				MaxAge:           "5h",
+			},
+			want: &CORSPolicy{
+				AllowCredentials: true,
+				AllowHeaders:     []string{"X-Some-Header-A", "X-Some-Header-B"},
+				AllowMethods:     []string{"GET", "PUT"},
+				AllowOrigin:      []CORSAllowOriginMatch{{Type: CORSAllowOriginMatchExact, Value: "*"}},
+				ExposeHeaders:    []string{"X-Expose-A", "X-Expose-B"},
+				MaxAge:           timeout.DurationSetting(5 * time.Hour),
+			},
+		},
+		"allow origin wildcard": {
+			cp: &contour_api_v1.CORSPolicy{
+				AllowMethods: []contour_api_v1.CORSHeaderValue{"GET"},
+				AllowOrigin:  []string{"*"},
+			},
+			want: &CORSPolicy{
+				AllowHeaders:  []string{},
+				AllowMethods:  []string{"GET"},
+				AllowOrigin:   []CORSAllowOriginMatch{{Type: CORSAllowOriginMatchExact, Value: "*"}},
+				ExposeHeaders: []string{},
+			},
+		},
+		"allow origin specific valid": {
+			cp: &contour_api_v1.CORSPolicy{
+				AllowMethods: []contour_api_v1.CORSHeaderValue{"GET"},
+				AllowOrigin:  []string{"http://foo-1.bar.com", "https://foo-2.com:443"},
+			},
+			want: &CORSPolicy{
+				AllowHeaders: []string{},
+				AllowMethods: []string{"GET"},
+				AllowOrigin: []CORSAllowOriginMatch{
+					{Type: CORSAllowOriginMatchExact, Value: "http://foo-1.bar.com"},
+					{Type: CORSAllowOriginMatchExact, Value: "https://foo-2.com:443"},
+				},
+				ExposeHeaders: []string{},
+			},
+		},
+		"allow origin invalid specific but valid regex": {
+			cp: &contour_api_v1.CORSPolicy{
+				AllowMethods: []contour_api_v1.CORSHeaderValue{"GET"},
+				AllowOrigin: []string{
+					"no-scheme.bar.com",
+					"http://bar.com/foo",
+					"http://baz.com?query1=2",
+					"http://example.org#fragment",
+				},
+			},
+			want: &CORSPolicy{
+				AllowHeaders: []string{},
+				AllowMethods: []string{"GET"},
+				AllowOrigin: []CORSAllowOriginMatch{
+					{Type: CORSAllowOriginMatchRegex, Value: "no-scheme.bar.com"},
+					{Type: CORSAllowOriginMatchRegex, Value: "http://bar.com/foo"},
+					{Type: CORSAllowOriginMatchRegex, Value: "http://baz.com?query1=2"},
+					{Type: CORSAllowOriginMatchRegex, Value: "http://example.org#fragment"},
+				},
+				ExposeHeaders: []string{},
+			},
+		},
+		"allow origin regex valid": {
+			cp: &contour_api_v1.CORSPolicy{
+				AllowMethods: []contour_api_v1.CORSHeaderValue{"GET"},
+				AllowOrigin:  []string{`.*\.foo\.com`, `https://example\.bar-[0-9]+\.org`},
```

Comment:
I'd say the format for wildcard domain matching in particular should be documented as an example in the godoc for AllowOrigin, since folks may think they can just put `*.foo.com` which is not quite right.